ouroboros-network/src/Ouroboros/Network/BlockFetch.hs

```diff
@@ -206,9 +210,45 @@ data BlockFetchConsensusInterface peer header block m =
        -- | Given a block header, validate the supposed corresponding block
        -- body.
        --
-       blockMatchesHeader      :: header -> block -> Bool
+       blockMatchesHeader      :: header -> block -> Bool,
+
+       -- | Calculate when a header's block was forged.
+       --
+       -- PRECONDITION: This function will succeed and give a _correct_ result
+       -- when applied to headers obtained via this interface (ie via
+       -- Consensus, ie via 'readCurrentChain' or 'readCandidateChains').
+       --
+       -- WARNING: This function may fail or, worse, __give an incorrect result
+       -- (!!)__ if applied to headers obtained from sources outside of this
+       -- interface. The 'FromConsensus' newtype wrapper is intended to make it
+       -- difficult to make that mistake, so please pay that syntactic price
+       -- and consider its meaning at each call to this function. Relatedly,
+       -- preserve that argument wrapper as much as possible when deriving
+       -- ancillary functions\/interfaces from this function.
+       headerForgeUTCTime :: FromConsensus header -> STM m UTCTime,
+
+       -- | Calculate when a block was forged.
+       --
+       -- PRECONDITION: Same as 'headerForgeUTCTime'.
+       --
+       -- WARNING: Same as 'headerForgeUTCTime'.
+       blockForgeUTCTime  :: FromConsensus block -> STM m UTCTime
      }
 
+-- | A new type used to emphasize the precondition of 'headerForgeUTCTime' and
+-- 'blockForgeUTCTime' at each call site.
+--
+-- At time of writing, the @a@ is either a header or a block. The headers are
+-- literally from Consensus (ie provided by ChainSync). Blocks, on the other
+-- hand, are indirectly from Consensus: they were fetched only because we
+-- favored the corresponding header that Consensus provided.
+newtype FromConsensus a = FromConsensus {unFromConsensus :: a}
+  deriving (Functor)
+
+instance Applicative FromConsensus where
+  pure = FromConsensus
+  FromConsensus f <*> FromConsensus a = FromConsensus (f a)
+
```

Comment:
You can't have the `FromConsensus` declaration here. The `BlockFetchConsensusInterface` is not used directly by blockfetch clients, they use `FetchClientPolicy`. The type needs to be visible from `ClientState.hs`.

Reply:
Thanks for clarifying, appreciate it. I've pushed a fix.